### DISCUSSION TITLE
override using @OVERRIDES@ not working within .withHeaders (Details below)

### DISCUSSION THREAD

**Discussion by mailtwogopal** [pactumjs/pactum#253] [General]
Hi,

I have a situation that any of my test needs a token as one of the request headers, and we rely on a Soap call (REST version not available yet) to get that token. In PactumJS, I managed to call that soap WS with .withBody(XMLrequest) and from the xml response received, I converted it to json response using **xml2js** node library. 

I was able to fetch the token from json response and using .returns stored that to a variable. When I try to override using @OVERRIDES@ inside .withHeaders like below, token value is set to undefined.

First call inside it block where am fetching token value from webservice1:

const fetchToken = await spec().withMethod('POST').withPath('url')
        .withJson(reqbody).withHeaders({ "test": "test", "Content-Type": "application/json" }).expectStatus(200)
        .returns((ctx) => {
            var xmlResp = ctx.res.body;
            convert.parseString(xmlResp, (err, jsonResp) => {
                if (err) console.log(err)
                tokenVal = jsonResp['test']['value'];
                return tokenVal;
            })
        });

Second call inside it block where am trying to use the const fetchToken inside my 2nd test like below but tokenvalue in header is set to undefined.
 
 await spec().withMethod('POST')
            .withPath('url').withHeaders({
                '@DATA:TEMPLATE@': 'headers',
                '@OVERRIDES@': {
                    'tokenvalue': `${fetchToken}` (also tried as simply as 'tokenvalue': fetchToken)
                }
            })
            .withJson({ '@DATA:TEMPLATE@': 'reqpayload' }).inspect()
            .expectStatus(200)

Can you check on this?

**Comment by mailtwogopal** [comment]
Please don't worry about small letters between the CAPS for DATA template etc. as am using it right in my tests

**Reply by ASaiAnudeep** [reply]
Instead of using dat templates, can you try with [Data Maps](https://pactumjs.github.io/guides/data-management.html#data-map)

```js
const { stash, spec } = require('pactum');

stash.addDataMap({
  'APP_TOKEN': 'some dynamic value'
});

await spec().post('<url>').withHeaders('tokenvalue', '$M{APP_TOKEN}')
```

**Reply by mailtwogopal** [reply]
Sure, I will give a try.. Thanks

**Reply by mailtwogopal** [reply]
Hi ASaiAnudeep,

I think the issue with below code where .returns does not return any value. This is how I modified to use DataMap as you suggested above.

stash.addDataMap({
            'appToken': 
            {
                'appTokenValue': await spec().withMethod('POST').withPath('URL')
                    .withJson({
                        "type": "test", "id": "test", "cred": "test"
                    }).withHeaders({ "Content-Type": "application/json" }).expectStatus(200)
                    .returns((ctx) => {
                        convert.parseString(ctx.res.body, (err, resp) => {
                            if (err) console.log(err)
                            return resp['Token']['value'];
                        })
                    }) 
        }})

and when I parse datamap like below 

const val = parse('$M{appToken.appTokenValue}')
console.log(val) is printing $M{appToken.appTokenValue}

Can you check if I am using .returns correct way, as inside it only I am converting xml response to json and fetch token value?

**Reply by mailtwogopal** [reply]
Hi ASaiAnudeep, This can be mark resolved as I was doing something wrong in returning inside async function. I realized that I had to pass callback function as the parameter to the function and then return token back with callback(token). I am all set and Thanks for your help here.
